### ISSUE TITLE
shim-15 for CentOS 7.7.1908

### ISSUE THREAD

**Issue by arrfab** [rhboot/shim-review#79]
Make sure you have provided the following information:

 - [x] link to your code branch cloned from rhboot/shim-review in the form user/repo@tag
https://github.com/CentOS/shim-review/tree/7.7.1908
 - [x] completed README.md file with the necessary information
https://github.com/CentOS/shim-review/blob/7.7.1908/README.md
 - [x] shim.efi to be signed : https://github.com/CentOS/shim-review/blob/7.7.1908/shimx64.efi
 - [x] any extra patches to shim via your own git tree or as files (see README.md)
 - [x] build logs : 
  - https://github.com/CentOS/shim-review/blob/7.7.1908/root.log
  - https://github.com/CentOS/shim-review/blob/7.7.1908/build.log

-------------------------------------------------------------------------------
What organization or people are asking to have this signed:
-------------------------------------------------------------------------------
Red Hat, Inc.

-------------------------------------------------------------------------------
What product or service is this for:
-------------------------------------------------------------------------------
CentOS Linux 7.7.1908

###### What is the origin and full version number of your shim?
shim 15 , pushed by Red Hat to git.centos.org

What's the justification that this really does need to be signed for the whole world to be able to boot it:
-------------------------------------------------------------------------------
CentOS Linux is deployed on a high number of nodes already using it in SecureBoot mode enabled

###### How do you manage and protect the keys used in your SHIM?
Can't discuss this publicly, but hardware HSM is used

###### Do you use EV certificates as embedded certificates in the SHIM?
Nope

###### What is the origin and full version number of your bootloader (GRUB or other)?
grub2, pushed by Red Hat to git.centos.org

###### If your SHIM launches any other components, please provide further details on what is launched
nothing else than grub2 and then kernel

###### How do the launched components prevent execution of unauthenticated code?
Our CI test shows that trying to boot an unsigned kernel isn't working, while a correct pesigned one works

###### Does your SHIM load any loaders that support loading unsigned kernels (e.g. GRUB)?
no

###### What kernel are you using? Which patches does it includes to enforce Secure Boot?
official RHEL 7.7 rebuild kernel.src.rpm

###### What changes were made since your SHIM was last signed?
RHEL 7.7 included new patches for shime 15, so we track those changes and rebuilt too

###### What is the hash of your final SHIM binary?
165919f84e88b815e7333e95ea6b2aa50d7f6f18e86bee43553f88bb2b43d05b  /boot/efi/EFI/centos/shimx64.efi

**Comment by arrfab** [comment]
Link to the Git repo holding all mandatory artifacts/build instructions : https://github.com/CentOS/shim-review/tree/7.7.1908

**Comment by avijc** [comment]
For what it's worth, I was able to reproduce the shim using the provided Dockerfile:
165919f84e88b815e7333e95ea6b2aa50d7f6f18e86bee43553f88bb2b43d05b  /build/builddir/BUILDROOT/shim-15-5.el7.centos.x86_64/usr/share/shim/x64-15-5.el7.centos/shimx64.efi

The Dockerfile essentially upgrades CentOS 7.6 to 7.7. Relevant installed packages are:

gcc-4.8.5-39.el7.x86_64
gnu-efi-3.0.8-2.el7.x86_64
gnu-efi-devel-3.0.8-2.el7.x86_64
make-3.82-24.el7.x86_64
redhat-rpm-config-9.1.0-88.el7.centos.noarch
rpm-build-4.11.3-40.el7.x86_64

Note that the (current) Dockerfile specifies redhat-rpm-config-9.1.0-80.el7 but that is no longer available (it was a CentOS 7.5.1804 package). However, that is not a problem because rpm-build pulls in redhat-rpm-config as a dependency, and this ends up installing the current -88 version.

**Comment by cyphermox** [comment]
Rebuild seems to go fine, but the sha256 is different between binaries, and yet there are no differences found in running hexdump.

```
Step 21/28 : RUN sha256sum /build/builddir/BUILDROOT/shim-15-5.el7.centos.x86_64/usr/share/shim/*-15-5.el7.centos/shim*.efi
 ---> Running in 8beb445e3f57
68b2763691a55825089ee7c85b0a17740daaa7d87e4964564c5d81b62806b838  /build/builddir/BUILDROOT/shim-15-5.el7.centos.x86_64/usr/share/shim/ia32-15-5.el7.centos/shimia32.efi
165919f84e88b815e7333e95ea6b2aa50d7f6f18e86bee43553f88bb2b43d05b  /build/builddir/BUILDROOT/shim-15-5.el7.centos.x86_64/usr/share/shim/x64-15-5.el7.centos/shimx64.efi
Removing intermediate container 8beb445e3f57
 ---> 78628b0aa6bd
Step 22/28 : COPY shimx64.efi /
 ---> 76ae25333d5f
Step 23/28 : RUN hexdump -Cv /build/builddir/BUILDROOT/shim-15-5.el7.centos.x86_64/usr/share/shim/x64-15-5.el7.centos/shimx64.efi > /built.hex
 ---> Running in 6d8f08bc18d4
Removing intermediate container 6d8f08bc18d4
 ---> be6e0ab84243
Step 24/28 : RUN hexdump -Cv /shimx64.efi > /orig.hex
 ---> Running in e8c923d2367a
Removing intermediate container e8c923d2367a
 ---> 54232cdf5314
Step 25/28 : RUN diff -u /orig.hex /built.hex  > /diff.hex
 ---> Running in fe82a0181f3a
Removing intermediate container fe82a0181f3a
 ---> 7d467e4a119b
Step 26/28 : RUN cat /diff.hex
 ---> Running in d6e026cf631c
Removing intermediate container d6e026cf631c
 ---> 3e70e21a97bb
Step 27/28 : RUN objdump -x /shimx64.efi | head -n 60
 ---> Running in 0c1016dcec71
```

Any ideas?

**Comment by arrfab** [comment]
@cyphermox : I can have a look but from your own log you have the same sha256sum ? 
```
165919f84e88b815e7333e95ea6b2aa50d7f6f18e86bee43553f88bb2b43d05b  /build/builddir/BUILDROOT/shim-15-5.el7.centos.x86_64/usr/share/shim/x64-15-5.el7.centos/shimx64.efi
```

And that's the one we ask a signing for : 165919f84e88b815e7333e95ea6b2aa50d7f6f18e86bee43553f88bb2b43d05b /boot/efi/EFI/centos/shimx64.efi
(we never asked the shimia32.efi, also present in your output log, to be signed)
Let me know if there is anything else different ?

**Comment by cyphermox** [comment]
No, it's fine, I was just confused by having two sha256 sums :)

This shim looks acceptable for signing.

**Comment by arrfab** [comment]
@cyphermox thanks a lot for the review.
@vathpela : let me know when it's submitted and then ping me back when you have it signed. We'll push it through "updates" repository (as we have released 7.7.1908 a long time ago now) :)
